OCastTests/TestOCastDomainGeneral.swift

```diff
@@ -0,0 +1,159 @@
+                "options": {
+                    "B": 2,
+                    "C": 3,
+                    "A": 1
+                },
+                "params": {
+                    "type": "type",
+                    "trackId": "trackId",
+                    "enable": true
+                }
+            }
+        },
+        "type": "type",
+        "src": "source",
+        "dst": "destination"
+    }
+    """.utf8)
+    
+    private let jsonRegexPattern = "\\{(\"status\":\"\\w+\",)?\"id\":\\d+,\"message\":\\{\"service\":\"\\w*\",\"data\":\\{(\"name\":\"\\w*\",)?(\"options\":\\{(\"\\w\":.+,?)+\\},)?\"params\":\\{(\"\\w*\":.*,?)*\\}\\}\\},\"type\":\"\\w*\",\"src\":\"\\w*\",\"dst\":\"\\w*\"\\}"
+    
+    private var jsonObject: OCastDeviceLayer<MediaTrackCommand>?
+    
+    override func setUp() {
+        
+        let mediaTrackCommand = MediaTrackCommand(type: "type", trackId: "trackId", enable: true)
+        let oCastDataLayer = OCastDataLayer(name: "name", params: mediaTrackCommand, options: ["A":1 , "B":2, "C":3])
+        let oCastApplicationLayer = OCastApplicationLayer(service: "service", data: oCastDataLayer)
+        let oCastDeviceLayer = OCastDeviceLayer(source: "source", destination: "destination", id: 1, status: "status", type: "type", message: oCastApplicationLayer)
+
+        jsonObject = oCastDeviceLayer
+    }
+    
+    override func tearDown() {
+    }
+    
+    func testDecodingWhenMissingId() throws {
+        assertThrowsKeyNotFound("id", decoding: OCastDeviceLayer<MediaTrackCommand>.self, from: try jsonData.json(deletingKeyPaths: "id"))
+    }
+    
+    func testDecodingWhenMissingMessage() throws {
+        assertThrowsKeyNotFound("message", decoding: OCastDeviceLayer<MediaTrackCommand>.self, from: try jsonData.json(deletingKeyPaths: "message"))
+    }
+    
+    func testDecodingWhenMissingType() throws {
+        assertThrowsKeyNotFound("type", decoding: OCastDeviceLayer<MediaTrackCommand>.self, from: try jsonData.json(deletingKeyPaths: "type"))
+    }
+    
+    func testDecodingWhenMissingDestination() throws {
+        assertThrowsKeyNotFound("dst", decoding: OCastDeviceLayer<MediaTrackCommand>.self, from: try jsonData.json(deletingKeyPaths: "dst"))
+    }
+    
+    func testDecodingWhenMissingSource() throws {
+        assertThrowsKeyNotFound("src", decoding: OCastDeviceLayer<MediaTrackCommand>.self, from: try jsonData.json(deletingKeyPaths: "src"))
+    }
+    
+    func testDecodingWhenWringValue() throws {
+    }
+    
+    func testEncoding() throws {
+        let jsonDataTest = try! JSONEncoder().encode(jsonObject)
+        let jsonStringTest = String(data: jsonDataTest, encoding: .utf8)!
+        let trimmedTest = jsonStringTest.replacingOccurrences(of: "\\s*", with: "", options: .regularExpression)
+        
+        XCTAssertNotNil(trimmedTest.range(of: jsonRegexPattern, options: .regularExpression, range: nil, locale: nil))
+    }
+    
+    func testDecoding() {
```

Comment:
each function must be tested for proper operation but also its case errors. You need to test a fail decoding.